### ISSUE TITLE
sourceFileMap not working when using WSL

### ISSUE THREAD

**Issue by trogne** [microsoft/vscode-cpptools#7389]
When debugging using gdb, on WSL ubuntu, I get this error : 

`Unable to open 'ioputs.c': Unable to read file 'vscode-remote://wsl+ubuntu/build/glibc-eX1tMB/glibc-2.31/libio/ioputs.c' (Error: Unable to resolve non-existing file 'vscode-remote://wsl+ubuntu/build/glibc-eX1tMB/glibc-2.31/libio/ioputs.c').`

It get this error even with sourceFileMap defined below.

I can make debugging work if I copy the full "/usr/src/glibc/glibc-2.31" directory into "/build/glibc-eX1tMB/".

So it seems that the "sourceFileMap" does not work here.

```
      {
        "name": "(gdb) Launch",
        "type": "cppdbg",
        "request": "launch",
        "program": "${workspaceFolder}//${fileBasenameNoExtension}",
        "stopAtEntry": false,
        "cwd": "${workspaceFolder}",
        "environment": [],
        "externalConsole": false,
        "MIMode": "gdb",
        "miDebuggerPath": "/usr/bin/gdb",
        "setupCommands": [
          {
            "description": "Enable pretty-printing for gdb",
            "text": "-enable-pretty-printing",
            "ignoreFailures": true
          }
        ],
        "sourceFileMap": {
          "/build/glibc-eX1tMB": "/usr/src/glibc"
        },
      },
```

- OS and Version: Windows 10
- VS Code Version: 1.55.2
- C/C++ Extension Version: v1.3.0

**Comment by WardenGnaw** [comment]
I believe this is an issue with vscode-remote since the debug adapter expects to find it in `vscode-remote://wsl+ubuntu/build/glibc-eX1tMB/` instead of just `/build/glibc-eX1tMB/`.

sourceFileMap expects the full path to the folder.

**Comment by weinand** [comment]
@WardenGnaw In the remote case VS Code accesses remote files via the `vscode-remote://wsl+ubuntu/...` URIs. So the use of URIs is expected and if those URIs cannot be resolved (remotely), then an error shows up.

The user said that copying the directory into the expected location makes the problem disappear. IMO this confirms that the problem is a broken "sourceFileMap".  

Since the "sourceFileMap" is implemented in the "cppdbg" extension, could you please verify that the map works in the remote case. For example it could that the mapping fails because is sees `vscode-remote://wsl+ubuntu/...` URIs that don't match the patterns.

**Comment by WardenGnaw** [comment]
@weinand It is looking for the path `vscode-remote://wsl+ubuntu/build/glibc-eX1tMB/glibc-2.31/` instead of  `/build/glibc-eX1tMB/`.

**Comment by weinand** [comment]
@WardenGnaw no, users should not have to deal with paths like "vscode-remote://wsl+ubuntu/..."
They are an implementation detail of the "remote extension host".
Debug adapters should not see those URIs because they are converted back to paths before they reach the DA.
So if the source mapping is done in the DA, it should work.
Where is the source mapping done?

**Comment by WardenGnaw** [comment]
This is done in the debug adapter, but the source file paths come from GDB.

The debug adapter is given `vscode-remote://wsl+ubuntu/build/glibc-eX1tMB/glibc-2.31/libio/ioputs.c` from the debugger (GDB) and we try to map that to the file the user wants. 

The debug adapter looks for anything that starts with on the left side of the sourceMap and will replace it with anything on the right side.

**Comment by weinand** [comment]
Why does the gdb debugger give "vscode-remote://wsl+ubuntu/..." URIs to the debug adapter?
That does not seem right. You should not see these URIs in the DA.

**Comment by WardenGnaw** [comment]
My statement about it coming from GDB is incorrect. GDB is giving the path without `vscode-remote://wsl+ubuntu/`

From my investigation, source file map works as expected. Here is the error without source maps:
![image](https://user-images.githubusercontent.com/3953714/115789755-7c333380-a37a-11eb-81a0-3a37504cda98.png)

Here is with the source map below:
```
"sourceFileMap":{
   "/build/glibc-YYA7BZ/": "/new/path/to/build/glibc-YYA7BZ/"
}
```
![image](https://user-images.githubusercontent.com/3953714/115789790-8b19e600-a37a-11eb-962a-ea11c474ab6e.png)

It should be properly prompting VS Code with the correct URI to open a file with the vscode-remote. 

@trogne Does the file `/usr/src/glibc/glibc-2.31/libio/ioputs.c` exist?

**Comment by weinand** [comment]
BTW, the local VS Code will send these URIs to the remote extension host which tries to resolve them against the (remote) file system. This functionality is independent from debugging, so the debug extension is not involved in serving these files.

**Comment by davidwin** [comment]
I think I'm hitting this issue with Remote SSH, but I'm confused whether we're supposed to have remote-vscode:// in the mapped-to part of `sourceFileMap` or not. If I put plain remote paths that map to my repository, gdb is able to make lookups to the correct source file when breaking on the first line of the application. However, I am not able to set any source breakpoint if I have this (`/repo/path/` being my repository where I set the breakpoint and `/some/alternate/path` is a path that may end up in the debug symbols due to shared ccache):
```json
"sourceFileMap": {
  "/some/alternate/path": "/repo/path"
}
```
Now, if I instead use `vscode-remote://ssh/repo/path` (not even specifyint the remote instance name), gdb is still able to break into the e.g. the first line of the application and show the correct source, _and_ my breakpoints work as well. Writing an invalid path after `vscode-remote://ssh/` breaks things again, so it seems like this is important.
So, is the solution to have "fully qualified" remote to-paths in `sourceFileMap`, then? I don't think it used to work this way a couple of releases back.

**Comment by beetaa** [comment]
Have the issue resolved? I am hitting the same issue, too. so confusing ...

![image](https://user-images.githubusercontent.com/5688224/132342544-81056478-3af1-4290-98a4-344befb094ae.png)

**Comment by i-am-david-fernandez** [comment]
For what it's worth, I seem to be having the same problem (or at least, the same symptom) trying to debug a python app in a container via remote-ssh (i.e., Windows VSCode -> Linux VM via remote-ssh -> python app in docker container).

I get this:

> Unable to open 'views.py': Unable to read file 'vscode-remote://ssh-remote+vm-hostname/a/b/health/views.py'
> (Error: Unable to resolve nonexistent file 'vscode-remote://ssh-remote+vm-hostname/a/b/health/views.py').

I realise I'm using a completely different debugger (`debugpy`), and a different remote mechanism (`ssh`) but the symptoms suggest a common root cause. I cannot get path mappings to have any effect (I've tried using both `pathMappings` and `sourceFileMap`), but if I create an appropriate symlink within the VM (such that the absolute path to `views.py` in the container matches that within the VM, via the symlink), then things seem to work fine.
